File: ios/Runner/PlaybackMemoryPolicy.swift
import Foundation

enum PlaybackMemoryPolicy {
  private static let formatterLock = NSLock()
  private static let fractionalFormatter: ISO8601DateFormatter = {
    let value = ISO8601DateFormatter()
    value.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    return value
  }()
  private static let secondsFormatter = ISO8601DateFormatter()
  private static let localFormatter: DateFormatter = {
    let value = DateFormatter()
    value.locale = Locale(identifier: "en_US_POSIX")
    value.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS"
    return value
  }()
  static func resume(positionMs: Int64, durationMs: Int64, progress: Double, completed: Bool) -> Int64 {
    if completed || positionMs < Int64(PlaybackPolicyValues.memoryResumeMinimumMs)
      || (durationMs > 0 && durationMs - positionMs <= Int64(PlaybackPolicyValues.memoryResumeRemainingMs))
      || progress >= Double(PlaybackPolicyValues.memoryCompletedPermille) / 1000 { return 0 }
    return positionMs
  }

  static func completed(positionMs: Int64, durationMs: Int64, progress: Double) -> Bool {
    if durationMs <= 0 { return progress >= Double(PlaybackPolicyValues.memoryUnknownCompletedPermille) / 1000 }
    return progress >= Double(PlaybackPolicyValues.memoryCompletedPermille) / 1000
      || durationMs - positionMs <= Int64(PlaybackPolicyValues.memoryCompletedRemainingMs)
  }

  static func timestamp(_ raw: String) -> Int64 {
    formatterLock.lock()
    defer { formatterLock.unlock() }
    let date = fractionalFormatter.date(from: raw)
      ?? secondsFormatter.date(from: raw) ?? localFormatter.date(from: raw)
    return Int64(((date?.timeIntervalSince1970 ?? 0) * 1000).rounded(.down))
  }

  static func nextTimestamp(now: String, existing: [String]) -> String {
    let next = existing.reduce(timestamp(now)) { max($0, timestamp($1) + 1) }
    return formatTimestamp(next)
  }

  static func formatTimestamp(_ milliseconds: Int64) -> String {
    formatterLock.lock()
    defer { formatterLock.unlock() }
    return fractionalFormatter.string(from: Date(timeIntervalSince1970: Double(milliseconds) / 1000))
  }
}

File: android/app/src/main/kotlin/com/example/starflow/PlaybackMemoryPolicy.kt
package com.example.starflow

import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

internal object PlaybackMemoryPolicy {
    fun resume(positionMs: Long, durationMs: Long, progress: Double, completed: Boolean): Long {
        if (completed || positionMs < PlaybackPolicyValues.memoryResumeMinimumMs ||
            (durationMs > 0 && durationMs - positionMs <= PlaybackPolicyValues.memoryResumeRemainingMs) ||
            progress >= PlaybackPolicyValues.memoryCompletedPermille / 1000.0) return 0
        return positionMs
    }

    fun completed(positionMs: Long, durationMs: Long, progress: Double): Boolean =
        if (durationMs <= 0) progress >= PlaybackPolicyValues.memoryUnknownCompletedPermille / 1000.0
        else progress >= PlaybackPolicyValues.memoryCompletedPermille / 1000.0 ||
            durationMs - positionMs <= PlaybackPolicyValues.memoryCompletedRemainingMs

    fun timestamp(raw: String): Long {
        val match = Regex("""^(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2})(?:\.(\d+))?(Z|[+-]\d{2}:?\d{2})?$""")
            .matchEntire(raw) ?: return 0
        val fraction = match.groupValues[2].padEnd(3, '0').take(3)
        val zone = match.groupValues[3]
        val format = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS" + if (zone.isEmpty()) "" else "Z", Locale.US)
        format.isLenient = false
        val suffix = if (zone == "Z") "+0000" else zone.replace(":", "")
        return runCatching { format.parse("${match.groupValues[1]}.$fraction$suffix")?.time ?: 0 }.getOrDefault(0)
    }

    fun nextTimestamp(now: String, existing: Iterable<String>): String {
        var next = timestamp(now)
        for (raw in existing) next = maxOf(next, timestamp(raw) + 1)
        return formatTimestamp(next)
    }

    fun formatTimestamp(value: Long): String = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }.format(Date(value))
}
